### PR TITLE
Add compatibility for Substrate v2.0.0-rc4

### PR DESCRIPTION
**Pull request by vgeddes**
Changes:
* Support for the `blake2b_128concat` hasher which used to construct storage keys.
* Support for the Extrinsic payload V4 format.
* Compatibility with changed `subkey inspect` output.

Reproducer for unsupported hash method (which this PR fixes):
```
git checkout master
export RPC_URL=ws://127.0.0.1:9944
export TEST_PRIV_KEY="//Alice"
/opt/go/bin/go test -timeout 30s github.com/Snowfork/go-substrate-rpc-client/teste2e -run ^TestChain_SubmitExtrinsic$
2020/07/24 23:54:06 Connecting to ws://127.0.0.1:9944...
--- FAIL: TestChain_SubmitExtrinsic (0.03s)
panic: hash function type blake2_128concat not yet supported [recovered]
        panic: hash function type blake2_128concat not yet supported
```

We are now able to submit extrinsics successfully:
```
/opt/go/bin/go test -timeout 30s github.com/Snowfork/go-substrate-rpc-client/teste2e -run ^TestChain_SubmitExtrinsic_Simple$
ok      github.com/Snowfork/go-substrate-rpc-client/teste2e     0.040s
```

### CODE REVIEW

File: signature/signature.go [hunk 1]
```diff
@@ -38,7 +38,7 @@ type KeyringPair struct {
 	PublicKey []byte
 }
 
-var rePubKey = regexp.MustCompile(`Public key \(hex\): 0x([a-f0-9]*)\n`)
+var rePubKey = regexp.MustCompile(`Public key \(hex\):\s+0x([a-f0-9]*)\n`)
```
Comment (philipstanislaus) [inline signature/signature.go:41]: :+1:

File: types/extrinsic_payload.go [hunk 2]
```diff
@@ -23,23 +23,24 @@ import (
 	"github.com/Snowfork/go-substrate-rpc-client/signature"
 )
 
-// ExtrinsicPayloadV3 is a signing payload for an Extrinsic. For the final encoding, it is variable length based on
+// ExtrinsicPayloadV4 is a signing payload for an Extrinsic. For the final encoding, it is variable length based on
```
Comment (philipstanislaus) [inline types/extrinsic_payload.go:26]: Instead of overwriting an older version, please just add the new one such that the library stays backwards compatible.

File: blake2b/blake2b.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+func (s *state) Write(p []byte) (n int, err error) {
+	// save the key for later use in Sum()
+	s.key = append(s.key, p...)
+
+	return s.hasher.Write(p)
+}
+
+func (s *state) Sum(b []byte) []byte {
+	// append key to final hash digest
+	return append(b, append(s.hasher.Sum(nil), s.key...)...)
```
Comment (philipstanislaus) [inline blake2b/blake2b.go:26]: Should `s.key` not exclusively be appended in the concat case?

File: blake2b/blake2b.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+func (s *state) Reset() {
+	s.hasher.Reset()
```
Comment (philipstanislaus) [inline blake2b/blake2b.go:30]: This should also reset `s.key`, right?

File: blake2b/blake2b.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+func (s *state) Size() int {
+	return s.hasher.Size()
```
Comment (philipstanislaus) [inline blake2b/blake2b.go:34]: This is not accurate, since `s.key` is appended to the `sum` of `hasher`.

```suggestion
	return s.hasher.Size() + len(key)
```

File: blake2b/blake2b.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+func (s *state) BlockSize() int {
+	return s.hasher.BlockSize()
+}
+
+func New128Concat(key []byte) (hash.Hash, error) {
```
Comment (philipstanislaus) [inline blake2b/blake2b.go:41]: Please add doc comments here

File: blake2b/blake2b.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+	hasher := state{
+		hasher: inner,
+		key:    key,
+	}
+
+	return &hasher, nil
+}
+
+func New128(key []byte) (hash.Hash, error) {
```
Comment (philipstanislaus) [inline blake2b/blake2b.go:55]: And here

File: blake2b/blake2b.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+	return blake2b.New(16, key)
+}
+
+func New256(key []byte) (hash.Hash, error) {
```
Comment (philipstanislaus) [inline blake2b/blake2b.go:59]: And here

File: blake2b/blake2b_test.go [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+func Test_128Concat(t *testing.T) {
+	key := []byte("abc")
+
+	h128, _ := New128(nil)
+	h128.Write(key)
+	h128Concat, _ := New128Concat(nil)
+	h128Concat.Write(key)
+
+	assert.Equal(t, append(h128.Sum(nil), key...), h128Concat.Sum(nil))
```
Comment (philipstanislaus) [inline blake2b/blake2b_test.go:18]: Please add tests for reset, size and blocksize.

File: blake2b/blake2b_test.go [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+	"testing"
+
+	. "github.com/Snowfork/go-substrate-rpc-client/blake2b"
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_128Concat(t *testing.T) {
+	key := []byte("abc")
+
+	h128, _ := New128(nil)
```
Comment (philipstanislaus) [inline blake2b/blake2b_test.go:13]: Please add the same tests that initializes with a key that is not nil to test whether initialization works as expected.

Comment (vgeddes) [reply blake2b/blake2b_test.go:13]: This file contains another test named `Test_128Concat_MAC`, where I've covered that case.

File: blake2b/blake2b.go (hunk 5 above)
Comment (vgeddes) [reply blake2b/blake2b.go:34]: Good catch, will push a fix.

File: blake2b/blake2b.go (hunk 3 above)
Comment (vgeddes) [reply blake2b/blake2b.go:26]: Well, I have to make sure to conform to the contract expected of `Sum` implementations, and ensure that the final checksum is appended to `b`.

As an aside, I'm actually changing the implementation to: `append(s.hasher.Sum(b), s.key...)` which is slightly more correct.

File: blake2b/blake2b.go (hunk 4 above)
Comment (vgeddes) [reply blake2b/blake2b.go:34]: fixed in latest upload

File: blake2b/blake2b.go (hunk 5 above)
Comment (vgeddes) [reply blake2b/blake2b.go]: fixed in latest upload

File: blake2b/blake2b.go (hunk 6 above)
Comment (vgeddes) [reply blake2b/blake2b.go:47]: fixed in latest upload

File: blake2b/blake2b.go (hunk 7 above)
Comment (vgeddes) [reply blake2b/blake2b.go:63]: fixed in latest upload

File: blake2b/blake2b.go (hunk 8 above)
Comment (vgeddes) [reply blake2b/blake2b.go:69]: fixed in latest upload

File: blake2b/blake2b_test.go (hunk 9 above)
Comment (vgeddes) [reply blake2b/blake2b_test.go]: done, added tests in the latest upload

File: types/extrinsic_payload.go (hunk 2 above)
Comment (vgeddes) [reply types/extrinsic_payload.go]: In the latest upload, I've gone half-way towards making that possible. I've renamed `types/extrinsic_payload.go` to `types/extrinsic_payloadV3.go` and have introduced `ExtrinsicPayloadV4` in `types/extrinsic_payloadV4.go`.

Ensuring GSRPC remains fully backward-compatible however, will require some further work. The core problem is that `extrinsic.go` would need to be refactored heavily to support backwards compatibility. For example, before this PR came about, `Extrinsic.Sign` assumed all payloads would be `ExtrinsicPayloadV3`.

Since we've forked the GSRPC repo, and this code is working well for us, I'd say there less urgency for us have full backwards compat. We could tackle the problem in a subsequent PR, or when we prepare to upstream our code into the Centrifuge repo. As it happens, there are other tasks we'd need to do before upstreaming, such as updating all the payload tests and exemplaries.

File: blake2b/blake2b_test.go [hunk 11]
```diff
@@ -12,10 +12,12 @@ func Test_128Concat(t *testing.T) {
 
 	h128, _ := New128(nil)
 	h128.Write(key)
+	h128.Write(key)
 	h128Concat, _ := New128Concat(nil)
 	h128Concat.Write(key)
+	h128Concat.Write(key)
 
-	assert.Equal(t, append(h128.Sum(nil), key...), h128Concat.Sum(nil))
+	assert.Equal(t, append(h128.Sum(nil), append(key, key...)...), h128Concat.Sum(nil))
```
Comment (philipstanislaus) [inline blake2b/blake2b_test.go:20]: Why do we append the key twice here?

File: blake2b/blake2b.go [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+package blake2b
+
+import (
+	"hash"
+
+	"golang.org/x/crypto/blake2b"
+)
+
+// State for the blake2b_128_concat hasher
+type state struct {
```
Comment (philipstanislaus) [inline blake2b/blake2b.go:12]: Please rename this for clarity:

```suggestion
type concatState struct {
```

File: blake2b/blake2b_test.go (hunk 11 above)
Comment (vgeddes) [reply blake2b/blake2b_test.go:20]: Because I wrote the key twice in the test. I wanted to test the that the new hashing scheme would work when `Write` was called multiple times.

File: blake2b/blake2b.go (hunk 12 above)
Comment (vgeddes) [reply blake2b/blake2b.go]: fixed in latest upload
